File: code/PuzzleSolver.java
import java.util.*;

public class PuzzleSolver {

    static Set<Integer> visited;




    // Coordinates of the objective position for each digit in the 8-puzzle
    static int[][] objectiveIndices = {{2, 2}, {0, 0}, {0, 1}, {0 ,2}, {1, 0}, {1, 1}, {1, 2}, {2, 0}, {2, 1}};
    static int[] allMoves = {3, 1, -1, -3};  // Possible moves: right, down, left, up
    static int[] movesMod3 = {3, -1, -3};    // Moves for the rightmost column
    static int[] movesMod3Plus1 = {3, 1, -3}; // Moves for the leftmost column
    static int GOAL = 123456780;  // Goal state of the puzzle












    // Function to calculate base^exponent in O(log(exponent))
    static int pow(int base, int exponent) {
        if (exponent == 0) return 1;
        if ((exponent & 1) == 1) return base * pow(base * base, exponent >> 1);
        return pow(base * base, exponent >> 1);
    }













    // Function to calculate Manhattan Distance heuristic
    static int manhattanDistance(int curr) {
        int man = 0;
        for (int i = 8; i > 0; --i) {
            int[] obj = objectiveIndices[(curr / pow(10, i)) % 10];
            man += Math.abs((8 - i) / 3 - obj[0]) + Math.abs((8 - i) % 3 - obj[1]);
        }
        return man;
    }














    // Structure to represent a node in the search space
    static class Node {
        List<Integer> path;
        int zeroIndex, cost, manhattan;

        Node(List<Integer> newPath, int lastMove, int newZeroIndex, int newCost) {
            path = new ArrayList<>(newPath);
            zeroIndex = newZeroIndex;
            cost = newCost;
            manhattan = manhattanDistance(lastMove);
            path.add(lastMove);
        }
    }















    // Function to swap two digits in the puzzle
    static int swapDigits(int curr, int i, int j) {
        i = 8 - i;
        j = 8 - j;
        int digitI = (curr / pow(10, i)) % 10;
        int digitJ = (curr / pow(10, j)) % 10;
        int newNum = curr - digitI * pow(10, i) - digitJ * pow(10, j);
        newNum += digitI * pow(10, j) + digitJ * pow(10, i);
        return newNum;
    }













    // Check if the new zero index is valid
    static boolean isValid(int zeroIndex) {
        return zeroIndex >= 0 && zeroIndex < 9;
    }














    // Choose set of moves based on the position of the zero
    static AbstractMap.SimpleEntry<int[], Integer> chooseMove(int zeroIndex) {
        if ((zeroIndex + 1) % 3 == 0) return new AbstractMap.SimpleEntry<>(movesMod3, 3);
        if (zeroIndex % 3 == 0) return new AbstractMap.SimpleEntry<>(movesMod3Plus1, 3);
        return new AbstractMap.SimpleEntry<>(allMoves, 4);
    }













    // Generate next possible moves from the current state
    static List<Node> nextMoves(List<Integer> currPath, int zeroIndex, int cost) {
        List<Node> successors = new ArrayList<>();
        cost++;
        AbstractMap.SimpleEntry<int[], Integer> moves = chooseMove(zeroIndex);

        for (int i = 0; i < moves.getValue(); i++) {
            int newZeroIndex = zeroIndex + moves.getKey()[i];

            if (isValid(newZeroIndex)) {
                int newMove = swapDigits(currPath.get(currPath.size() - 1), newZeroIndex, zeroIndex);
                if (!visited.contains(newMove)) {
                    successors.add(new Node(currPath, newMove, newZeroIndex, cost));
                    visited.add(newMove);
                }
            }
        }
        return successors;
    }













    // Find the position of the zero in the puzzle
    static int findZero(int start, int i) {
        if (start % 10 == 0) return i;
        return findZero(start / 10, i - 1);
    }













    // Custom comparator for the priority queue in A* search
    static class NodeComparator implements Comparator<Node> {
        public int compare(Node p1, Node p2) {
            return Integer.compare(p1.cost + p1.manhattan, p2.cost + p2.manhattan);
        }
    }














    // Function to solve the 8-puzzle using A*
    static List<Integer> puzzleSolver(int start) {
        PriorityQueue<Node> pq = new PriorityQueue<>(new NodeComparator());
        visited = new HashSet<>();

        pq.add(new Node(new ArrayList<>(), start, findZero(start, 8), 0));

        while (!pq.isEmpty()) {
            Node curr = pq.poll();

            int currLast = curr.path.get(curr.path.size() - 1);

            if (currLast == GOAL) return curr.path;

            for (Node child : nextMoves(curr.path, curr.zeroIndex, curr.cost)) {
                pq.add(child);
            }
        }
        return new ArrayList<>();
    }













    // Main function to test the puzzle solver
    public static void main(String[] args){

        List<Integer> result = puzzleSolver(385742160);

        for (int i : result) System.out.println(i);
        
        System.out.println("\nLength of the solution path: " + result.size());

    }
}

File: code/PuzzleSolver.py
from queue import PriorityQueue

# Coordinates of the objective position for each digit in the 8-puzzle
objective_indices = [[2, 2], [0, 0], [0, 1], [0, 2], [1, 0], [1, 1], [1, 2], [2, 0], [2, 1]]
all_moves = [3, 1, -1, -3]  # Possible moves: right, down, left, up
moves_mod_3 = [3, -1, -3]    # Moves for the rightmost column
moves_mod_3_plus_1 = [3, 1, -3]  # Moves for the leftmost column
GOAL = 123456780  # Goal state of the puzzle








# Function to calculate base^exponent in O(log(exponent))
def power(base, exponent):
    if not exponent:
        return 1
    if exponent & 1:
        return base * power(base * base, exponent >> 1)
    return power(base * base, exponent >> 1)








# Function to calculate Manhattan Distance heuristic
def manhattan_distance(curr):
    man = 0
    for i in range(8, 0, -1):
        obj = objective_indices[(curr // power(10, i)) % 10]
        man += abs((8 - i) // 3 - obj[0]) + abs((8 - i) % 3 - obj[1])
    return man





        
        
        
        
class Node:
    
    def __init__(self, path, last_move, zero_index, cost):
        self.path = path
        self.zero_index = zero_index
        self.cost = cost
        self.manhattan = manhattan_distance(last_move)
        self.path.append(last_move)

    def __lt__(self, other):
        # Define the comparison for the priority queue
        return (self.cost + self.manhattan) < (other.cost + other.manhattan)














# Function to swap two digits in the puzzle
def swap_digits(curr, i, j):
    i = 8 - i
    j = 8 - j
    digit_i = (curr // power(10, i)) % 10
    digit_j = (curr // power(10, j)) % 10
    new_num = curr - digit_i * power(10, i) - digit_j * power(10, j)
    new_num += digit_i * power(10, j) + digit_j * power(10, i)
    return new_num













# Check if the new zero index is valid
def is_valid(zero_index):
    return 0 <= zero_index < 9












# Choose set of moves based on the position of the zero
def choose_move(zero_index):
    if (zero_index + 1) % 3 == 0:
        return moves_mod_3
    if zero_index % 3 == 0:
        return moves_mod_3_plus_1
    return all_moves















# Generate next possible moves from the current state
def next_moves(curr_path, zero_index, cost):
    successors = []
    cost += 1
    moves = choose_move(zero_index)

    for i in moves:
        new_zero_index = zero_index + i

        if is_valid(new_zero_index):
            new_move = swap_digits(curr_path[-1], new_zero_index, zero_index)
            if new_move not in visited:
                successors.append(Node(curr_path.copy(), new_move, new_zero_index, cost))
                visited.add(new_move)

    return successors

















# Find the position of the zero in the puzzle
def find_zero(start, i=8):
    if start % 10 == 0:
        return i
    return find_zero(start // 10, i - 1)















# Function to solve the 8-puzzle using A*
def puzzle_solver(start):
    
    
    pq = PriorityQueue()
    pq.put(Node([], start, find_zero(start), 0))

    while not pq.empty():
        curr = pq.get()
        curr_last = curr.path[-1]

        if curr_last == GOAL:
            return curr.path

        for child in next_moves(curr.path, curr.zero_index, curr.cost):
            pq.put(child)

    return []












visited = set()
result = puzzle_solver(385742160)
for i in result:
    print(i)
print("\nLength of the solution path:", len(result))
